app/models/enterprise.rb

```diff
@@ -117,7 +117,7 @@ class Enterprise < ActiveRecord::Base
     # checkout.
     ready_enterprises = Enterprise.ready_for_checkout.select('enterprises.id')
     if ready_enterprises.present?
-      where("enterprises.id NOT IN (?)", ready_enterprises)
+      where("enterprises.id NOT IN (?)", ready_enterprises.map(&:id))
```

Comment:
I think that we can use `pluck(:id)` here to be more efficient. The `select` isn't needed then. But this could have a performance impact, the array can be big.

Reply:
pluck could create an extra query, right?
I think this should be select :-)
https://github.com/openfoodfoundation/openfoodnetwork/pull/3709
I think we should agree on some rules for using map vs select vs pluck :-)

Reply:
> I think we should agree on some rules for using map vs select vs pluck :-)

That sounds good! And we could add it to the wiki

Reply:
Thank you for the clarification. I now think that I know how it works.

In this particular case, there was a select that would lead to a sub-query. The `present?` call creates a separate query just looking if anything is there without loading the enterprises. Then applying `map` would first load all the enterprise objects and then copy the ids into an array. A `pluck` call would do that without creating the enterprise objects (less Ruby work).

If we used the enterprise objects somewhere else, we should use map to re-use the same array and not load the enterprises again. But I don't think we do it here, so pluck would be better than map.

Matt, you explained that `select *` is a problem with sub-queries in Rails 4. That's logical. We should definitely avoid that. In this case, we do `select('enterprises.id')` in line 118 which should allow us to do an efficient sub-query. That should be no problem with Rails 4, right?

With my current knowledge, we don't need `map` here and should use a sub-query. We could move the `select` from line 118 down to the 120 to make it clearer that we are selecting the ids for the `IN` query part.

Reply:
awesome Maikel 👌

Reply:
I've done a bit more reading on Rails subqueries. The docs seem to suggest we need to remove both the `select` _and_ the `pluck`, and leave Rails to build the subquery..? 

I'll have a play with it in the Rails 4 branch and see if I can improve things without Postgres complaining.

Reply:
There's some other interesting errors when playing with this in the Rails 4 branch, including: `ActiveRecord::StatementInvalid: PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters, but prepared statement "" requires 1`.

Some relevant discussion is here: https://github.com/rails/rails/issues/20077

Apparently there are some other issues with subqueries using associations / collections / prepared statements. Adding `#to_sql` on some of the pieces seems to work in some cases, eg: `distributor_id: user.enterprises.to_sql`

In this specific case, the nice Rails 4 way is to use the `#not` method to avoid the prepared statement, with something like: `where.not(id: ready_enterprises)`, but we can't do that in `master`.

Reply:
Aha! Yeah it looks like it's mostly the prepared statements in our scopes that are causing issues. In the `#managed_by` scope in  `enterprise_fee.rb` this line throws a "subquery has too many columns" error:

```ruby
where('enterprise_id IN (?)', user.enterprises)
```

but this does not: 

```ruby
where(enterprise_id: user.enterprises)
```

Reply:
nice investigation Matt!!
is there a reason you are not satisfied with how we have been fixing things so far with:
where('enterprise_id IN (?)', user.enterprises.select(&:id))

Reply:
I need to redo the other subquery PR and check some things in Rails 4. It looks like it will be better if we avoid prepared statements and favour the hash syntax, passing an `ActiveRecord::Relation` object (as mentioned above). And in that case I think we generally don't need the `select`, as `ActiveRecord` should in theory figure it out and build the subquery correctly?

Reply:
I am not sure I agree.
Prepared statements are a good thing in general. I dont think we should plan to just avoid them.
I dont think we should let active record control things, we should control things and we should name the exact fields we want in the query.
These are general rules I have seen used in other projects: as much prepared statements as possible and lists of fields in a query as specific as possible.

Reply:
In general, the new Rails 4 way `where.not(id: ready_enterprises)` sounds better than what we have at the moment. But for compatibility with Rails 3 and 4, I would prefer `      where("enterprises.id NOT IN (?)", ready_enterprises.select("enterprises.id"))` for now. The reason I prefer this version is that it's very explicit what's happening, it's compatible with many versions and avoids the Rails 3 bug where multiple `where(id: collection)` statements override each other.

Instead of optimising the way we use the ActiveRecord interface further, I would also suggest looking at the underlying problem and changing our logic here. This negation is really expensive. It would be better to find criteria for enterprises that make them not being ready for checkout and using those in a direct query.

One problem is that there can be several reasons why an enterprise is not ready. And we are querying them all together. Instead, we could actually query those individually and change the UI to say:

- These enterprises are missing a shipping method: ...
- These enterprises are missing a payment method: ...
- ...

@Matt-Yorkley I noticed that your most recent version still uses `pluck`. Does `select('enterprises.id')` still raise the too-many-column error? I don't understand that because `enterprises.id` is only one column, isn't it? Or are there many more in JOIN clauses that count as well?